### PR TITLE
Adding text editor functionality to text viewer.

### PR DESCRIPTION
**Pull request by nevumx**
Pursuant to my feature request https://github.com/d0k3/GodMode9/issues/861, I am pleased to present my proposal PR for adding text editor functionality to the existing text viewer by repurposing the existing keyboard that is used for file renames.

Features:
- UTF-8 display and typing support
- Support for both LF and CRLF line endings, not CR line endings
- PGUP/PGDN/HOME/END-style cursor functionality
- Full wordwrapped/non-wordwrapped text support

Screenshots:
snap_240627192950
snap_240627161019
snap_240627161030
snap_240627161239
snap_240627161245

Considerations:
- Some functions shared by, for example, the README ToC Viewer and System information viewer has been changed.
  - I tested the system information viewer with no issues
  - The README viewer seems to have disappeared for some reason, so I couldn't test that... 🤔
- In order to fully support editing both wordwrapped and non-wordwrapped text, there were some edge cases that resulted in some unsightly `if` statements around, for example, when the last line is a certain length.
- I tried to get the cursor to blink first by using code like `InputWait(1)`, but that resulted in inputs getting dropped occasionally when pressing a button between `InputWait` calls when the screen was being drawn.
  - I also tried a multi-threaded approach, but even though the code I wrote with `<threads.h>` compiled, the thread would never spawn due to [`thrd_error`](https://en.cppreference.com/w/c/thread/thrd_errors)... I am guessing threads are out of scope for this project, since I don't see them anywhere else, definitely open to other suggestions on this...
- I renamed the "Textviewer" to "Text Editor," which might look kind of odd next to "Hexeditor;" but "Texteditor" seemed weird too, so I am open to suggestions here too...
- Some hardware buttons used on the rename prompt as keyboard hotkeys are different than in the text editor.
  - For example, "R" doesn't switch between capslock states, instead it is used to move the cursor further, like with the text viewer.
  - Instead "Y" switches between capslock states, because I don't see a purpose for a hardware button to insert a space into the text...
  - "A" Adds a newline char instead of submitting.

Definitely open to sugestions on this, especially as it relates to key bindings/text/etc.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -116,7 +116,7 @@ With the possibilites GodMode9 provides, not everything may be obvious at first
 * __Search drives and folders__: Just press R+A on the drive / folder you want to search.
 * __Compare and verify files__: Press the A button on the first file, select `Calculate SHA-256`. Do the same for the second file. If the two files are identical, you will get a message about them being identical. On the SDCARD drive (`0:`) you can also write an SHA file, so you can check for any modifications at a later point.
 * __Hexview and hexedit any file__: Press the A button on a file and select `Show in Hexeditor`. A button again enables edit mode, hold the A button and press arrow buttons to edit bytes. You will get an additional confirmation prompt to take over changes. Take note that for certain files, write permissions can't be enabled.
-* __View text files in a text viewer__: Press the A button on a file and select `Show in Textviewer` (only shows up for actual text files). You can enable wordwrapped mode via R+Y, and navigate around the file via R+X and the dpad.
```
Comment (nevumx) [inline README.md:119]: Open to "Text Editor"/"Texteditor"/"Hex Editor"/"Hexeditor" guidance here...

File: arm9/source/common/swkbd.h [hunk 2]
```diff
@@ -53,9 +65,9 @@ enum {
     KEY_ALPHA, ' ', KEY_BKSPC
 
 #define SWKBD_KEYS_NUMPAD \
-    '7', '8', '9', 'F', 'E', \
```
Comment (nevumx) [inline arm9/source/common/swkbd.h:56]: I changed the arrangement of these to be more consistent with e.g. calculators.

File: arm9/source/common/swkbd.h [hunk 3]
```diff
@@ -68,12 +80,20 @@ enum {
...
+    12, 0, \
+    8, 32, 85, 32, 32, 18, 18, 18, 18, 0, \
+    0
+
 #define SWKBD_LAYOUT_SPECIAL \
     6, 0, \
     6, 0, \
     6, 0, \
     6, 0, \
-    3, 32, 46, 32, 0, \
```
Comment (nevumx) [inline arm9/source/common/swkbd.h:76]: This was bothering me as well.

File: arm9/source/common/swkbd.h [hunk 4]
```diff
@@ -87,3 +107,5 @@ enum {
 
 #define ShowKeyboardOrPrompt (TouchIsCalibrated() ? ShowKeyboard : ShowStringPrompt)
 bool PRINTF_ARGS(3) ShowKeyboard(char* inputstr, u32 max_size, const char *format, ...);
+bool BuildKeyboard(TouchBox* swkbd, const char* keys, const u8* layout, bool multi_line);
```
Comment (nevumx) [inline arm9/source/common/swkbd.h:110]: This is exposed to prevent rebuilds between keypresses in the `MemTextViewer` function.

File: arm9/source/godmode.c [hunk 5]
```diff
@@ -1225,7 +1225,7 @@ u32 FileHandlerMenu(char* current_path, u32* cursor, u32* scroll, PaneData** pan
     int n_opt = 0;
     int special = (special_opt) ? ++n_opt : -1;
     int hexviewer = ++n_opt;
-    int textviewer = (filetype & TXT_GENERIC) ? ++n_opt : -1;
```
Comment (nevumx) [inline arm9/source/godmode.c:1228]: Allowing for editing of empty files.

File: arm9/source/godmode.c [hunk 6]
```diff
@@ -1316,6 +1316,7 @@ u32 FileHandlerMenu(char* current_path, u32* cursor, u32* scroll, PaneData** pan
     }
     else if (user_select == textviewer) { // -> show in text viewer
         FileTextViewer(file_path, scriptable);
+        GetDirContents(current_dir, current_path);
```
Comment (nevumx) [inline arm9/source/godmode.c:1319]: This is necessary, I assume, to recalculate file sizes?

File: arm9/source/utils/scripting.c [hunk 7]
```diff
@@ -259,86 +265,125 @@ static inline u32 hexntostr(const u8* hex, char* str, u32 len) {
     return len;
 }
 
-static inline u32 line_len(const char* text, u32 len, u32 ww, const char* line, char** eol) {
-    u32 last = ((text + len) - line);
+static inline bool is_crlf(const char* str) {
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:268]: Majority rule seems to make the most sense here... We _could_ do a hotkey to switch newline types on the keyboard, and display the current newline type status on the keyboard screen too, let me know if you think that's a good idea.

File: arm9/source/utils/scripting.c [hunk 8]
```diff
@@ -259,86 +265,125 @@ static inline u32 hexntostr(const u8* hex, char* str, u32 len) {
...
+    do if (str[0] == '\n') ++lf; else if (str[0] == '\r' && str[1] == '\n') ++crlf, ++str;
+    while (*str++);
+    return crlf > lf;
+}
+
+static inline bool is_newline(const char* chr) {
+    return chr[0] == '\n' || (chr[0] == '\r' && chr[1] == '\n');
+}
+
+static inline u32 bytes_in_chars_u32(const char* str, u32 nchars) {
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:279]: These are for slightly saner multibyte character support.

File: arm9/source/utils/scripting.c [hunk 9]
```diff
@@ -259,86 +265,125 @@ static inline u32 hexntostr(const u8* hex, char* str, u32 len) {
...
+    return chars;
+}
+
+static inline u32 chars_between_pointers(const char* start, const char* end) {
+    u32 chars = 0;
+    for (const char* i = start; *i && i < end; IncChar(&i), ++chars);
+    return chars;
+}
+
+static inline u32 line_len_chars(const char* text, u32 len, u32 ww, const char* line, const char** eol) {
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:303]: I made these a bit more simplified while supporting multibyte chars, but as a result, their behavior has changed slightly. I changed invocations accordingly where necessary.

File: arm9/source/utils/scripting.c [hunk 10]
```diff
@@ -259,86 +265,125 @@ static inline u32 hexntostr(const u8* hex, char* str, u32 len) {
...
                 add = 0;
             }
         }
 
-
         return l0;
     }
 }
 
+static inline const char* line_start(const char* text, u32 len, u32 ww, const char* ptr) {
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:383]: For example, this can be used now instead of `line_seek(..., 0);`

File: arm9/source/utils/scripting.c [hunk 11]
```diff
@@ -1614,7 +1659,7 @@ bool run_line(const char* line_start, const char* line_end, u32* flags, char* er
 
 // checks for illegal ASCII symbols
 bool ValidateText(const char* text, u32 len) {
-    if (!len) return false;
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:1617]: Allow editing empty files.

File: arm9/source/utils/scripting.c [hunk 12]
```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
...
-    u32 p_ar = TV_LLEN_DISP - strlen(ar_str);
-    u32 x_al = x_txt + (p_al * FONT_WIDTH_EXT);
-    u32 x_ar = x_txt + (p_ar * FONT_WIDTH_EXT);
+    u32 x_al = x_txt;
+    u32 x_ar = x_txt + ((TV_LLEN_DISP - strlen(ar_str)) * FONT_WIDTH_EXT);
 
     // display text on screen
-    char txtstr[TV_LLEN_DISP + 1];
-    char* ptr = line0;
+    char txtstr[TV_LLEN_DISP * MAX_CHAR_SIZE + 1];
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:1680]: Add UTF-8 char max size.

File: arm9/source/utils/scripting.c [hunk 13]
```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
...
     u32 nln = lno;
+    bool last_empty_line_drawn = false;
     for (u32 y = TV_VPAD; y < SCREEN_HEIGHT; y += FONT_HEIGHT_EXT + (2*TV_VPAD)) {
-        char* ptr_next = line_seek(text, len, ww, ptr, 1);
-        u32 llen = line_len(text, len, ww, ptr, NULL);
-        u32 ncpy = ((int) llen < off_disp) ? 0 : (llen - off_disp);
-        if (ncpy > TV_LLEN_DISP) ncpy = TV_LLEN_DISP;
-        bool al = !ww && off_disp && (ptr != ptr_next);
-        bool ar = !ww && (llen > off_disp + TV_LLEN_DISP);
+        int off_disp_bytes = bytes_in_chars_int(ptr, off_disp_chars);
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:1685]: Multibyte support.

File: arm9/source/utils/scripting.c [hunk 14]
```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
...
+            memmove(txtstr + strlen(al_str), txtstr + txtstr_al_bytes, sizeof(txtstr) - txtstr_al_bytes);
+            memcpy(txtstr, al_str, strlen(al_str));
+        }
 
         // draw line number & text
         DrawString(TOP_SCREEN, txtstr, x_txt, y, color_text, COLOR_STD_BG);
         if (TV_LNOS > 0) { // line number
-            if (ptr != ptr_next)
-                DrawStringF(TOP_SCREEN, x_lno, y, ((ptr == text) || (*(ptr-1) == '\n')) ? COLOR_TVOFFS : COLOR_TVOFFSL, COLOR_STD_BG, "%0*lu", TV_LNOS, nln);
+            bool prev_ww_line_full = ww && ww == chars_between_pointers(line_seek_chars(text, len, ww, ptr, -1), ptr);
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:1730]: Unfortunate edge case to draw line number when the `'\0'` is word-wrapped onto its own line...

File: arm9/source/utils/scripting.c [hunk 15]
```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
...
+                if (last_line_empty) last_empty_line_drawn = true;
+            }
             else DrawStringF(TOP_SCREEN, x_lno, y, COLOR_TVOFFSL, COLOR_STD_BG, "%*.*s", TV_LNOS, TV_LNOS, " ");
         }
 
+        if (cursor) {
+            u32 cursor_line_offset_chars = chars_between_pointers(ptr + off_disp_bytes, cursor);
+            if (cursor >= ptr + off_disp_bytes && cursor <= ptr + off_disp_bytes + ncpy_bytes && cursor_line_offset_chars < TV_LLEN_DISP
+                    && (cursor != ptr + off_disp_bytes + ncpy_bytes || is_newline(cursor) || cursor == text + len)) {
+                DrawRectangle(TOP_SCREEN, x_txt + cursor_line_offset_chars * FONT_WIDTH_EXT, y, FONT_WIDTH_EXT, 1, COLOR_RED);
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:1743]: Without a blinking line, a hollow red rect is the next best cursor I could think of. These coords are picked so that they are drawn over by text.

File: arm9/source/utils/scripting.c [hunk 16]
```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
...
+                DrawRectangle(TOP_SCREEN, x_txt + (cursor_line_offset_chars + 1) * FONT_WIDTH_EXT - ((cursor_line_offset_chars == TV_LLEN_DISP - 1) ? 1 : 0), y, 1, FONT_HEIGHT_EXT, COLOR_RED);
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:1744]: Small edge case to squish the cursor box at the end of the line/screen.

File: arm9/source/utils/scripting.c [hunk 17]
```diff
@@ -1710,56 +1813,181 @@ bool MemTextViewer(const char* text, u32 len, u32 start, bool as_script) {
...
+                const char* cursor_line_start = line_start(text, len, ww, cursor);
+                u32 cursor_chars_from_line_start = chars_between_pointers(cursor_line_start, cursor);
+                if (cursor_chars_from_line_start < off_disp_chars + strlen(al_str)) off_disp_chars = cursor_chars_from_line_start - strlen(al_str);
+                if (cursor_chars_from_line_start >= off_disp_chars + TV_LLEN_DISP - strlen(ar_str)) off_disp_chars = cursor_chars_from_line_start + strlen(ar_str) - TV_LLEN_DISP + 1;
+            }
+            while (cursor && cursor < line0_next) line0_next = line_seek_chars(text, len, ww, line0_next, -1);
+            while (cursor && line0_next < line_seek_chars(text, len, ww, GetNextChar(cursor), -TV_NLIN_DISP)) line0_next = line_seek_chars(text, len, ww, line0_next, 1);
+        }
+
+        // find last allowed lines (ww and nonww)
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:1952]: These checks have to be performed every iteration now.

File: arm9/source/utils/scripting.c [hunk 18]
```diff
@@ -1710,56 +1813,181 @@ bool MemTextViewer(const char* text, u32 len, u32 start, bool as_script) {
...
+        if (off_disp_chars + TV_LLEN_DISP > llen_max) off_disp_chars = llen_max - TV_LLEN_DISP;
+        if (off_disp_chars < 0 || ww) off_disp_chars = 0;
+    }
+
+    // check for user edits
+    if (text_cpy) {
+        if (save_path) {
+            bool diffs = false;
+            if (len != text_cpy_len) diffs = true;
+            else for (u32 i = 0; i < len; ++i) if (text[i] != text_cpy[i]) { diffs = true; break; }
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:1985]: This could be slow for large files, but I took it form the hex editor; open to suggestions here...

**Comment by nevumx** [comment]
Not exactly sure who to tag here... @Wolfvak / @d0k3 ? Anyone else?

File: arm9/source/utils/scripting.c [hunk 19]
```diff
@@ -1846,18 +2074,20 @@ bool MemToCViewer(const char* text, u32 len, const char* title) {
 bool FileTextViewer(const char* path, bool as_script) {
     // load text file (completely into memory)
     // text file needs to fit inside the STD_BUFFER_SIZE
-    u32 flen, len;
+    size_t fileSize = FileGetSize(path);
+    if (fileSize >= STD_BUFFER_SIZE) {
```
Comment (nevumx) [inline arm9/source/utils/scripting.c:2078]: Max file size check. I tested with a much lower number as well to make sure that the editor would not allow characters to be typed beyond the max file size.

**Comment by Trickiy** [comment]
Could the hex editing functionality be added to the hex viewer as well?

**Comment by nevumx** [comment]
> Could the hex editing functionality be added to the hex viewer as well?

The hex viewer has had edit functionality since 2016: https://github.com/d0k3/GodMode9/commit/88a62d8f2ecc50f1ee99e0a08359ae85d02dd0f3

**Comment by nevumx** [comment]
FYI @d0k3 / @Wolfvak I have added clipboard select/cut/copy/paste support as well. The description has been updated!

File: arm9/source/common/swkbd.h (hunk 4 above)
Comment (BlueRaja) [reply arm9/source/common/swkbd.h:111]: A lot of these Github comments should probably be code comments instead

Comment (nevumx) [reply arm9/source/common/swkbd.h:113]: @BlueRaja done. I went through and added code comments where I felt they were relevant. If you have any other concerns, please let me know. 😁

**Comment by nevumx** [comment]
@d0k3 / @Wolfvak, any more thoughts on this PR? I think text editing would be a nice feature to have, for example, for editing config files directly in GM9 without the need for more software.

**Comment by nevumx** [comment]
@d0k3 / @Wolfvak @BlueRaja, any more thoughts on this PR? I think text editing would be a nice feature to have, for example, for editing config files directly in GM9 without the need for more software.

**Comment by nevumx** [comment]
@d0k3 / @Wolfvak / @BlueRaja Bumping this

**Comment by d0k3** [comment]
@nevumx - sorry about the long radio silence, know that your contribution is appreciated. As you see, I didn't merge this p/r for the current release, however that doesn't mean it won't be in a future release. Could you rebase this on the current master branch?

File: arm9/source/lua/gm9ui.c [hunk 20]
```diff
@@ -206,12 +206,16 @@ static int ui_show_text_viewer(lua_State* L) {
 
     // validate text ourselves so we can return a better error
     // MemTextViewer calls ShowPrompt if it's bad, and i don't want that
+
+    if (len >= STD_BUFFER_SIZE) {
```
Comment (nevumx) [inline arm9/source/lua/gm9ui.c:210]: Adding this to be more in line with `FileTextViewer(...)` code.

File: arm9/source/lua/gm9ui.c [hunk 21]
```diff
@@ -225,6 +229,11 @@ static int ui_show_file_text_viewer(lua_State* L) {
     // validate text ourselves so we can return a better error
     // MemTextViewer calls ShowPrompt if it's bad, and i don't want that
     // and FileTextViewer calls the above function
+
+    size_t fileSize = FileGetSize(path);
```
Comment (nevumx) [inline arm9/source/lua/gm9ui.c:233]: Adding this to be more in line with `FileTextViewer(...)` code.

File: arm9/source/lua/gm9ui.c [hunk 22]
```diff
@@ -233,16 +242,14 @@ static int ui_show_file_text_viewer(lua_State* L) {
 
     // TODO: replace this with something that can detect file read errors and actual 0-length files
     size_t flen = FileGetData(path, text, STD_BUFFER_SIZE - 1, 0);
-
     text[flen] = '\0';
-    u32 len = (ptrdiff_t)memchr(text, '\0', flen + 1) - (ptrdiff_t)text;
```
Comment (nevumx) [inline arm9/source/lua/gm9ui.c:238]: This line is unnecessary; using `flen` instead as we do in the `FileTextViewer(...)` code.

**Comment by nevumx** [comment]
> @nevumx - sorry about the long radio silence, know that your contribution is appreciated. As you see, I didn't merge this p/r for the current release, however that doesn't mean it won't be in a future release. Could you rebase this on the current master branch?

@d0k3 No problem, I rebased, and fixed the Lua invocations of `MemTextViewer(...)` to be more in line with the `FileTextViewer(...)` code. I did a quick smoke test on my 3DS to make sure everything was working well, which it was. I don't know how to test Lua though, so I would just give that a quick spin before merging. That being said, I can't imagine anything being much different with the `MemTextViewer(...)` calls I see in the Lua code.

**Comment by nevumx** [comment]
@d0k3 How is the outlook for getting this merged? 😅
